js/src/logger.ts

```diff
@@ -2924,6 +3131,78 @@ function validateAndSanitizeExperimentLogPartialArgs(
   }
 }
 
+/**
+ * Creates a deep copy of the given event. Replaces references to user objects
+ * with placeholder strings to ensure serializability, except for
+ * {@link Attachment} objects, which are preserved.
+ */
+function deepCopyEvent<T extends Partial<BackgroundLogEvent>>(event: T): T {
+  // We both check for serializability and round-trip `partialRecord` through
+  // JSON in order to create a "deep copy". This has the benefit of cutting
+  // out any reference to user objects when the object is logged
+  // asynchronously, so that in case the objects are modified, the logging is
+  // unaffected. In the future, this could be changed to a "real" deep copy so
+  // that immutable types (large strings) do not have to be copied.
+
+  const attachments: Attachment[] = [];
+  const IDENTIFIER = "_bt_saved_attachment";
+
+  const serialized = JSON.stringify(event, (_k, v) => {
+    if (v instanceof SpanImpl) {
+      return `<span>`;
+    } else if (v instanceof Experiment) {
+      return `<experiment>`;
+    } else if (v instanceof Dataset) {
+      return `<dataset>`;
+    } else if (v instanceof Logger) {
+      return `<logger>`;
+    } else if (v instanceof Attachment) {
+      const idx = attachments.push(v);
+      return { [IDENTIFIER]: idx - 1 };
+    }
+    return v;
+  });
+  const x = JSON.parse(serialized, (_k, v) => {
+    if (
+      v instanceof Object &&
+      IDENTIFIER in v &&
+      typeof v[IDENTIFIER] === "number"
+    ) {
+      return attachments[v[IDENTIFIER]];
+    }
+    return v;
+  });
+  return x;
+}
+
+/**
+ * Helper function for uploading attachments. Recursively extracts `Attachment`
+ * values and replaces them with their `AttachmentReferences`.
+ *
+ * @param value The event to filter. Will be modified in-place.
+ * @param attachments Flat array of extracted attachments.
+ */
+function extractAttachments(value: any, attachments: Attachment[]): void {
```

Comment:
The type of `value` could be `Record<string, unknown>`?

Reply:
I agree that this will allow us to make the signature use `Record<string, unknown>`. I'm not sure what we gain though other than making the type checker happy. The input is already round-trip serialized anyway. We can guard that `Object.entries` will work by checking `value instanceof Object`. In case of `Date` or `Function`, `Object.entries` returns empty array so the recursive call is a no-op. In case of `Array`, Object.entries iterates the array so we are also covered.

On the other hand, we need to pay the cost of zod's deep copy at each level of recursion for an event which is potentially large.

Reply:
> We can guard that `Object.entries` will work by checking `value instanceof Object`. In case of `Date` or `Function`, `Object.entries` returns empty array so the recursive call is a no-op. In case of `Array`, Object.entries iterates the array so we are also covered.

Unit test will cover these cases now: 41280add25df748c1c0c1be4e7db5dec9d0cdb87